### PR TITLE
Music library share operations

### PR DESCRIPTION
**Pull request by pwt**
This is a preliminary PR for functionality that partially addresses https://github.com/SoCo/SoCo/issues/678.

I'm not expecting merge approval at this stage. Instead, I'm seeking feedback on the implementation itself, and also on what else is required to reach done/done. Tests, documentation, etc.

Note that I've hand-cooked the XML extraction in the `MusicLibrary` class, for the `list_library_shares()` method, and it returns a list of shares as strings. I suspect this approach is non-standard. Happy to have pointers on how to standardise this.

### CODE REVIEW

**Comment by KennethNielsen** [comment]
Hi @pwt 

I think this looks like great and I can't wait to see the add method. I will drop a few comments inline, but it's all details.

**Comment by KennethNielsen** [comment]
Besides from the comments which are all minor you asked about additional requirements.

We unfortunately never got into a position where we had full test coverage and we therefore also do not require it of new additions, (it would be somewhat hypocritical, since you can probably find some of my early code untested, from before I adopted a passion for software tests). That being said we would of course very much appreciate it. Tests are in the test folder, run with pytest and there should be unit tests for this module in the `test_music_library.py`file. So if you feel up for it, unit tests documenting different cases would be greatly appreciated. But you are also welcome to push it for a later PR if you prefer to get this in first.

As for documentation we don't require additional documentation besides that which is generated from the doc strings. But if you feel like writing some, I think this addition would be a prime candidate for a section in the micro examples: http://docs.python-soco.com/en/latest/examples.html.

BTW. If you have questions about tools etc., besides from this PR, there is also our newly formed gitter channel: https://gitter.im/SoCo/SoCo

File: soco/music_library.py [hunk 1]
```diff
@@ -597,3 +598,47 @@ def album_artist_display_option(self):
...
+        """
+        response = self.contentDirectory.Browse([
+            ('ObjectID', 'S:'),
+            ('BrowseFlag', 'BrowseDirectChildren'),
+            ('Filter', '*'),
+            ('StartingIndex', '0'),
+            ('RequestedCount', '100'),
+            ('SortCriteria', '')
+        ])
+        xml_dict = xmltodict.parse(response['Result'])
```
[the remark this answers is not in the document]
Comment (pwt) [reply soco/music_library.py]: Good catch. For now I think I'll stick with the dictionary approach and catch KeyError exceptions. I'll test against a Sonos system with no shares to check the validity of the the approach.

File: soco/music_library.py [hunk 2]
```diff
@@ -597,3 +598,47 @@ def album_artist_display_option(self):
...
+    def delete_library_share(self, share_name):
+        """Delete a music library share.
+
+        Args:
+            share_name (str): the name of the share to be deleted
+            share_name should be of the form `//hostname_or_IP/share_path`
+            e.g., `//my_computer.local/my_music/lossless`
+
+        Returns:
+            0 on success, 1 on failure
```
[the remark this answers is not in the document]
Comment (pwt) [reply soco/music_library.py]: I've removed any return value, and I'll document that the call raises a SoCoUPnPException on failure.

File: soco/music_library.py [hunk 3]
```diff
@@ -597,3 +598,47 @@ def album_artist_display_option(self):
         """
         result = self.contentDirectory.GetAlbumArtistDisplayOption()
         return result['AlbumArtistDisplayOption']
+
+    def list_library_shares(self):
+        """List the music library shares.
```
[the remark this answers is not in the document]
Comment (pwt) [reply soco/music_library.py]: Done.

File: soco/music_library.py [hunk 4]
```diff
@@ -597,3 +598,47 @@ def album_artist_display_option(self):
...
+
+        Returns:
+            A list of shares
```
[the remark this answers is not in the document]
Comment (pwt) [reply soco/music_library.py]: Corrected.

**Comment by pwt** [comment]
Update: I've now tested `music_library.list_library_shares()` against a Sonos system with no shares present, and it's working fine.

**Comment by pwt** [comment]
I think I'm done/done now on listing and deleting shares. Adding shares is significantly harder, and may not be possible, so I propose leaving that for a potential future PR.

**Comment by KennethNielsen** [comment]
> Adding shares is significantly harder, and may not be possible, so I propose leaving that for a potential future PR.

Good idea.

File: soco/music_library.py [hunk 5]
```diff
@@ -597,3 +598,49 @@ def album_artist_display_option(self):
...
+        if matches == '1':
+            shares.append(unpacked['dc:title'])
+            return shares
+        # Otherwise it's multiple matches
+        for share in unpacked:
+            shares.append(share['dc:title'])
+        return shares
+
+    def delete_library_share(self, share_name):
+        """Delete a music library share. This is a 'fire and forget' operation
```
Comment (KennethNielsen) [inline soco/music_library.py]: You really don't need the "fire and forget" warning, and if you choose to keep it, I would suggest a neutral metaphor ;)

File: tests/test_music_library.py [hunk 6]
```diff
@@ -156,3 +156,58 @@ def test_soco_start_library_update(self, moco):
         moco.contentDirectory.RefreshShareIndex.assert_called_with([
             ('AlbumArtistDisplayOption', ''),
         ])
+
+    def test_soco_list_library_shares(self, moco):
```
Comment (KennethNielsen) [inline tests/test_music_library.py]: Excellent

**Comment by KennethNielsen** [comment]
I tested the list function, works great.

File: doc/examples.rst [hunk 7]
```diff
@@ -160,3 +160,27 @@ Clearing the queue is done with the
...
+Listing and deleting music library shares
+-----------------------------------------
+
+Music library shares are the local network drive shares connected to
+Sonos, which host the audio content in the Sonos Music Library.
+
+To list the shares connected to Sonos, use the
+:meth:`~soco.music_library.MusicLibrary.list_library_shares` method as follows::
+
+  >>> library = soco.music_library.MusicLibrary()
```
[the remark this answers is not in the document]
Comment (pwt) [reply doc/examples.rst]: Got it. Thanks.

File: doc/examples.rst [hunk 8]
```diff
@@ -160,3 +160,27 @@ Clearing the queue is done with the
...
+
+The result is a list of network share locations.
+
+To delete a network share, use the
+:meth:`~soco.music_library.MusicLibrary.delete_library_share` method as follows::
+
+  >>> library = soco.music_library.MusicLibrary()
+  >>> library.delete_library_share('//share_host_01/music')
+
+This method is 'fire and forget', so you might want to confirm removal
```
[the remark this answers is not in the document]
Comment (pwt) [reply doc/examples.rst]: The reason I included the warning text is that I've had instances when the operation succeeds, but Sonos does not remove the share. Still, I guess people will figure this out from themselves.

File: soco/music_library.py (hunk 5 above)
Comment (pwt) [reply soco/music_library.py]: Ok.

File: tests/test_music_library.py (hunk 6 above)
Comment (pwt) [reply tests/test_music_library.py]: Thanks.

File: doc/examples.rst [hunk 9]
```diff
@@ -160,3 +160,27 @@ Clearing the queue is done with the
...
+  >>> library = soco.music_library.MusicLibrary()
+  >>> library.list_library_shares()
+  ['//share_host_01/music', '//share_host_02/music']
+
+The result is a list of network share locations.
+
+To delete a network share, use the
+:meth:`~soco.music_library.MusicLibrary.delete_library_share` method as follows::
+
+  >>> library = soco.music_library.MusicLibrary()
```
[the remark this answers is not in the document]
Comment (pwt) [reply doc/examples.rst]: Ok.

**Comment by pwt** [comment]
Thanks for guidance and comments. No hurry for the review.

File: doc/examples.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (KennethNielsen) [reply doc/examples.rst]: Ah. If you actually have experience that this happens, then I agree that we should warn. Then maybe just reformulate in a neutral form.

[the remark this answers is not in the document]
Comment (pwt) [reply doc/examples.rst]: I'm in two minds about this, but I'll include some (more 'woke') text.

**Comment by KennethNielsen** [comment]
All checks out. Merged.

Thanks for the contribution! Please add a comment to the release notes thread #674 about the change.

**Comment by pwt** [comment]
Thanks for merging the PR!

FWIW, I'd recommend using **squash merge** in future, to avoid messy intermediate commits on the main branch: https://github.blog/2016-04-01-squash-your-commits/

**Comment by KennethNielsen** [comment]
I read up on it. Thanks.

**Comment by KennethNielsen** [comment]
Thanks for the hint on the squash commits. I wasn't aware that this option had been added. I think whether to use them or not on an individual commit vs. asking the contributor for an interactive rebase, would depend on the size of the PR. I think I will open a discussion about this.

**Comment by pwt** [comment]
I agree -- I think it should be taken on a case-by-case basis. Some cases will be obvious candidates for squash merge, in others there may be value in the commit history.

Thanks for considering it.

**Pull request merged**